src/transformStateWithClones.js

```diff
@@ -6,8 +6,36 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const allStates = [];
+  let stateCopy = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        stateCopy = Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
+        break;
+      default:
+        break;
+    }
+
+    allStates.push(Object.assign({}, stateCopy));
```

Comment:
```suggestion
    allStates.push({ ...stateCopy });
```

It's not a mistake, but usually we do copy using this variant, it looks simplier 😉